### PR TITLE
chore: feat `format` for `configuration.properties` `biome.lspBin`

### PR DESCRIPTION
**Pull request by Malix-off**
https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema:~:text=does%20not%20match.-,format,-for%20restricting%20strings

### CODE REVIEW

**Comment by nhedger** [comment]
This seems to clash with the default empty string.

Screenshot 2024-01-29 at 09 02 20

**Comment by nhedger** [comment]
Additionally, it seems to require a scheme, which would be a breaking change for most.

The following wouldn't work anymore, for example:

- `./biome`
- `/some/path/with/biome`

**Comment by Malix-off** [comment]
Weird, imma dig into it, as it theorically is a valid URI

**Review by Malix-off — COMMENTED** [review]
@nhedger

I just discovered that [the vscode api documentation about contributions points schema property](https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema:~:text=Other%20JSON%20Schema%20properties) is lacking **A LOT** of properties and documentation.

This is a list of all the properties the VScode LSP is showing containing the letter `"a"` in it.
image

There is clearly a ton missing.
I will report it to the VScode team. [_(update)_](https://github.com/biomejs/biome-vscode/pull/106#ref-issue-2106075848)

**Comment by Malix-off** [comment]
`Developer: inspect context keys` result of `biome.lspBin` UI:
```json
{
    "config.breadcrumbs.enabled": true,
    "config.debug.toolBarLocation": "floating",
    "config.editor.minimap.enabled": true,
    "config.editor.renderControlCharacters": true,
    "config.editor.renderWhitespace": "selection",
    "config.editor.stickyScroll.enabled": true,
    "config.extensions.autoUpdate": true,
    "config.files.autoSave": "onFocusChange",
    "config.git.allowForcePush": true,
    "config.git.enabled": true,
    "config.notebook.globalToolbar": true,
    "config.notebook.stickyScroll.enabled": true,
    "config.security.workspace.trust.enabled": true,
    "config.terminal.integrated.stickyScroll.enabled": true,
    "config.terminal.integrated.tabs.enabled": true,
    "config.terminal.integrated.tabs.showActions": "singleTerminalOrNarrow",
    "config.terminal.integrated.tabs.showActiveTerminal": "singleTerminalOrNarrow",
    "config.window.menuBarVisibility": "classic",
    "config.window.titleBarStyle": "custom",
    "config.workbench.activityBar.location": "side",
    "config.workbench.editor.editorActionsLocation": "default",
    "config.workbench.editor.enablePreview": true,
    "config.workbench.editor.showTabs": "single",
    "config.workbench.layoutControl.type": "both",
    "config.workbench.sideBar.location": "right",
    "config.workbench.statusBar.visible": true,
    "isComposing": false,
    "refactorPreview.defaultViewContainerLocation": true,
    "refactorPreview.defaultViewLocation": true,
    "refactorPreview.canMove": false,
    "workbench.panel.repl.defaultViewContainerLocation": true,
    "workbench.views.service.panel.087ed677-adbe-48e8-a104-a4f66a87d1b8.defaultViewContainerLocation": true,
    "workbench.panel.repl.view.canMove": true,
    "workbench.panel.markers.defaultViewContainerLocation": true,
    "workbench.panel.markers.view.defaultViewLocation": true,
    "workbench.panel.markers.view.active": true,
    "workbench.panel.markers.view.visible": true,
    "workbench.panel.markers.view.canMove": true,
    "workbench.panel.output.defaultViewContainerLocation": true,
    "workbench.panel.output.canMove": true,
    "workbench.panel.testResults.defaultViewContainerLocation": true,
    "workbench.panel.testResults.view.defaultViewLocation": true,
    "workbench.panel.testResults.view.canMove": true,
    "terminal.defaultViewContainerLocation": true,
    "terminal.canMove": true,
    "workbench.view.debug.defaultViewContainerLocation": true,
    "workbench.debug.variablesView.defaultViewLocation": true,
    "workbench.debug.watchExpressionsView.defaultViewLocation": true,
    "workbench.debug.callStackView.defaultViewLocation": true,
    "workbench.debug.breakPointsView.defaultViewLocation": true,
    "workbench.debug.welcome.defaultViewLocation": true,
    "workbench.debug.loadedScriptsView.defaultViewLocation": true,
    "workbench.debug.variablesView.canMove": true,
    "workbench.debug.watchExpressionsView.canMove": true,
    "workbench.debug.callStackView.canMove": true,
    "workbench.debug.breakPointsView.canMove": true,
    "workbench.debug.welcome.canMove": false,
    "workbench.debug.loadedScriptsView.canMove": true,
    "workbench.view.testing.defaultViewLocation": false,
    "workbench.view.testing.canMove": true,
    "workbench.view.explorer.defaultViewContainerLocation": true,
    "outline.defaultViewLocation": true,
    "timeline.defaultViewLocation": true,
    "outline.active": true,
    "outline.visible": true,
    "outline.canMove": true,
    "timeline.canMove": true,
    "workbench.view.remote.defaultViewContainerLocation": true,
    "workbench.view.scm.defaultViewContainerLocation": true,
    "workbench.scm.defaultViewLocation": true,
    "workbench.scm.repositories.defaultViewLocation": true,
    "workbench.scm.active": true,
    "workbench.scm.visible": true,
    "workbench.scm.canMove": true,
    "workbench.scm.repositories.canMove": true,
    "workbench.view.extension.test.defaultViewContainerLocation": true,
    "workbench.view.testCoverage.defaultViewLocation": true,
    "workbench.view.testCoverage.canMove": true,
    "workbench.view.extensions.defaultViewContainerLocation": true,
    "workbench.view.search.defaultViewContainerLocation": true,
    "workbench.view.search.defaultViewLocation": true,
    "workbench.view.search.active": true,
    "workbench.view.search.visible": true,
    "workbench.view.search.canMove": true,
    "viewContainer.refactorPreview.enabled": false,
    "viewContainer.workbench.panel.repl.enabled": false,
    "viewContainer.workbench.panel.markers.enabled": true,
    "viewContainer.workbench.panel.output.enabled": false,
    "viewContainer.workbench.panel.testResults.enabled": false,
    "viewContainer.terminal.enabled": false,
    "terminal.defaultViewLocation": false,
    "workbench.panel.output.defaultViewLocation": false,
    "workbench.panel.repl.view.defaultViewLocation": false,
    "terminal.active": true,
    "workbench.panel.output.active": true,
    "terminal.visible": true,
    "workbench.panel.output.visible": true,
    "viewContainer.workbench.views.service.panel.087ed677-adbe-48e8-a104-a4f66a87d1b8.enabled": true,
    "viewContainer.workbench.view.debug.enabled": true,
    "viewContainer.workbench.view.explorer.enabled": true,
    "viewContainer.workbench.view.remote.enabled": false,
    "viewContainer.workbench.view.scm.enabled": true,
    "viewContainer.workbench.view.extension.test.enabled": false,
    "viewContainer.workbench.view.extensions.enabled": true,
    "viewContainer.workbench.view.search.enabled": true,
    "focusedView": "",
    "navigableContainerFocused": true,
    "listSupportsTypeNavigation": true,
    "listSupportsKeyboardNavigation": true,
    "canNavigateBack": false,
    "canNavigateForward": false,
    "canNavigateBackInNavigationLocations": false,
    "canNavigateForwardInNavigationLocations": false,
    "canNavigateToLastNavigationLocation": false,
    "canNavigateBackInEditLocations": false,
    "canNavigateForwardInEditLocations": false,
    "canNavigateToLastEditLocation": false,
    "canReopenClosedEditor": true,
    "autoSaveAfterShortDelayContext": false,
    "activePanel": "",
    "panelFocus": false,
    "activeViewlet": "workbench.view.explorer",
    "sideBarFocus": false,
    "activeAuxiliary": "",
    "auxiliaryBarFocus": false,
    "saveConflictResolutionContext": false,
    "workbench.panel.chatSidebar.defaultViewContainerLocation": true,
    "viewContainer.workbench.panel.chatSidebar.enabled": false,
    "remoteConnectionState": "",
    "inTerminalRunCommandPicker": false,
    "terminalSuggestWidgetVisible": false,
    "canToggleWordWrap": false,
    "editorWordWrap": false,
    "wslFeatureInstalled": true,
    "currentProfile": "__default__profile__",
    "profiles.enabled": true,
    "isCurrentProfileTransient": false,
    "hasProfiles": false,
    "accessibilityModeEnabled": false,
    "timelineHasProvider": true,
    "timeline.active": true,
    "refactorPreview.enabled": false,
    "isMac": false,
    "isLinux": false,
    "isWindows": true,
    "isWeb": false,
    "isMacNative": false,
    "isIOS": false,
    "isMobile": false,
    "remoteName": "",
    "virtualWorkspace": "",
    "temporaryWorkspace": false,
    "hasWebFileSystemAccess": true,
    "isDevelopment": true,
    "productQualityType": "stable",
    "activeEditor": "workbench.editor.settings2",
    "activeEditorIsReadonly": true,
    "activeEditorCanToggleReadonly": false,
    "activeEditorCanRevert": true,
    "activeEditorCanSplitInGroup": false,
    "activeEditorAvailableEditorIds": "",
    "editorIsOpen": true,
    "textCompareEditorVisible": false,
    "textCompareEditorActive": false,
    "sideBySideEditorActive": false,
    "activeEditorGroupEmpty": false,
    "activeEditorGroupIndex": 1,
    "activeEditorGroupLast": true,
    "activeEditorGroupLocked": false,
    "multipleEditorGroups": false,
    "dirtyWorkingCopies": false,
    "inputFocus": false,
    "workbenchState": "empty",
    "workspaceFolderCount": 0,
    "openFolderWorkspaceSupport": true,
    "emptyWorkspaceSupport": true,
    "enterMultiRootWorkspaceSupport": true,
    "splitEditorsVertically": false,
    "isFullscreen": false,
    "isAuxiliaryWindowFocusedContext": false,
    "inZenMode": false,
    "isCenteredLayout": false,
    "mainEditorAreaVisible": true,
    "editorTabsVisible": false,
    "sideBarVisible": true,
    "titleBarVisible": true,
    "titleBarStyle": "custom",
    "panelPosition": "bottom",
    "panelVisible": false,
    "panelMaximized": false,
    "panelAlignment": "center",
    "auxiliaryBarVisible": false,
    "inQuickOpen": false,
    "syncEnabled": true,
    "syncStatus": "idle",
    "userDataSyncAccountStatus": "available",
    "enableSyncActivityViews": false,
    "hasConflicts": false,
    "enableSyncConflictsView": false,
    "hasOutdatedExtensions": false,
    "notificationCenterVisible": false,
    "notificationToastsVisible": false,
    "explorerViewletVisible": true,
    "outlineFollowsCursor": false,
    "outlineFiltersOnType": true,
    "outlineSortMode": 0,
    "outlineAllCollapsed": false,
    "view.outline.visible": false,
    "workbench.explorer.openEditorsView.defaultViewLocation": true,
    "workbench.explorer.emptyView.defaultViewLocation": true,
    "workbench.explorer.openEditorsView.active": true,
    "workbench.explorer.emptyView.active": true,
    "workbench.explorer.openEditorsView.visible": true,
    "workbench.explorer.emptyView.visible": true,
    "groupFocusedInOpenEditors": false,
    "dirtyEditorFocusedInOpenEditors": false,
    "readonlyEditorFocusedInOpenEditors": true,
    "view.workbench.explorer.openEditorsView.visible": true,
    "view.workbench.explorer.emptyView.visible": true,
    "workbench.explorer.openEditorsView.canMove": true,
    "workbench.explorer.emptyView.canMove": false,
    "workbench.view.extension.references-view.defaultViewContainerLocation": false,
    "viewContainer.workbench.view.extension.references-view.enabled": false,
    "jsBrowserBreakpoints.defaultViewLocation": true,
    "jsExcludedCallers.defaultViewLocation": true,
    "jsBrowserBreakpoints.canMove": true,
    "jsExcludedCallers.canMove": true,
    "npm.defaultViewLocation": false,
    "npm.canMove": true,
    "references-view.tree.defaultViewLocation": true,
    "references-view.tree.canMove": true,
    "forwardedPortsViewEnabled": true,
    "activeCustomEditorId": "",
    "focusedCustomEditorIsEditable": false,
    "terminalWebExtensionContributedProfile": false,
    "terminalGroupCount": 0,
    "terminalEditorActive": false,
    "terminalIsOpen": false,
    "terminalProcessSupported": true,
    "terminalHasBeenCreated": false,
    "terminalCount": 0,
    "isProfileExportInProgress": false,
    "isProfileImportInProgress": false,
    "userDataProfiles.defaultViewContainerLocation": true,
    "viewContainer.userDataProfiles.enabled": false,
    "activeWebviewPanelId": "",
    "debuggersAvailable": true,
    "debugExtensionAvailable": true,
    "expressionSelected": false,
    "loadedScriptsSupported": false,
    "stepBackSupported": false,
    "focusedSessionIsAttach": false,
    "focusedSessionIsNoDebug": false,
    "restartFrameSupported": false,
    "stepIntoTargetsSupported": false,
    "jumpToCursorSupported": false,
    "debugSetVariableSupported": false,
    "debugSetExpressionSupported": false,
    "multiSessionDebug": false,
    "terminateDebuggeeSupported": false,
    "suspendDebuggeeSupported": false,
    "disassembleRequestSupported": false,
    "focusedStackFrameHasInstructionReference": false,
    "interactiveWindowOpen": false,
    "testing.isRunning": false,
    "testing.hasAnyResults": false,
    "activeOutputChannel": "tasks",
    "activeLogOutput": false,
    "workspaceHasCommenting": false,
    "debugState": "inactive",
    "hasDebugged": false,
    "inDebugMode": false,
    "debugUx": "simple",
    "breakpointsExist": false,
    "disassemblyViewFocus": false,
    "workbench.debug.welcome.active": true,
    "workbench.debug.welcome.visible": true,
    "taskRunning": false,
    "serverlessWebContext": false,
    "customExecutionSupported": true,
    "taskCommandsRegistered": true,
    "shellExecutionSupported": true,
    "processExecutionSupported": true,
    "hasChatProvider": false,
    "scm.providerCount": 0,
    "scmRepositorySortKey": "discoveryTime",
    "shareProviderCount": 0,
    "hasSpeechProvider": false,
    "testing.hasRunnableTests": false,
    "testing.hasDebuggableTests": false,
    "testing.hasCoverableTests": false,
    "testing.hasNonDefaultProfile": false,
    "testing.hasConfigurableProfile": false,
    "testing.supportsContinuousRun": false,
    "testing.providerCount": 0,
    "testing.canRefresh": false,
    "testing.isRefreshing": false,
    "testing.activeEditorHasTests": false,
    "workbench.panel.repl.view.active": true,
    "workbench.panel.repl.view.visible": true,
    "inlineChatHasProvider": false,
    "git.closedRepositoryCount": 0,
    "git.parentRepositoryCount": 0,
    "git.unsafeRepositoryCount": 0,
    "git.state": "initialized",
    "gitOpenRepositoryCount": "0",
    "gitVersion2.35": true,
    "git-base.gitEnabled": true,
    "multiCursorModifier": "altKey",
    "userHasOpenedNotebook": true,
    "logLevel": "info",
    "remoteTunnelConnection": "disconnected",
    "isIssueTroubleshootActive": false,
    "releaseNotesUrl": "https://go.microsoft.com/fwlink/?LinkID=533483#vscode",
    "downloadUrl": "https://code.visualstudio.com",
    "updateState": "idle",
    "majorMinorUpdateAvailable": false,
    "~remote.forwardedPortsContainer.defaultViewContainerLocation": false,
    "viewContainer.~remote.forwardedPortsContainer.enabled": false,
    "~remote.forwardedPorts.defaultViewLocation": false,
    "~remote.forwardedPorts.active": true,
    "~remote.forwardedPorts.visible": true,
    "~remote.forwardedPorts.canMove": true,
    "scmActiveResourceHasChanges": false,
    "userDataSyncTurningOn": false,
    "workbench.view.sync.defaultViewContainerLocation": true,
    "viewContainer.workbench.view.sync.enabled": false,
    "workbench.views.sync.conflicts.defaultViewLocation": true,
    "workbench.views.sync.conflicts.canMove": false,
    "treeView.workbench.views.sync.remoteActivity.enableCollapseAll": true,
    "treeView.workbench.views.sync.remoteActivity.enableRefresh": true,
    "workbench.views.sync.remoteActivity.defaultViewLocation": true,
    "workbench.views.sync.remoteActivity.canMove": false,
    "treeView.workbench.views.sync.machines.enableRefresh": true,
    "workbench.views.sync.machines.defaultViewLocation": true,
    "workbench.views.sync.machines.canMove": false,
    "treeView.workbench.views.sync.localActivity.enableCollapseAll": true,
    "treeView.workbench.views.sync.localActivity.enableRefresh": true,
    "workbench.views.sync.localActivity.defaultViewLocation": true,
    "workbench.views.sync.localActivity.canMove": false,
    "treeView.workbench.views.sync.troubleshoot.enableRefresh": true,
    "workbench.views.sync.troubleshoot.defaultViewLocation": true,
    "workbench.views.sync.troubleshoot.canMove": false,
    "treeView.workbench.views.sync.externalActivity.enableCollapseAll": false,
    "treeView.workbench.views.sync.externalActivity.enableRefresh": false,
    "workbench.views.sync.externalActivity.defaultViewLocation": true,
    "workbench.views.sync.externalActivity.canMove": false,
    "editSessionsShowView": false,
    "editSessionsPending": false,
    "editSessionsSignedIn": true,
    "workbench.view.editSessions.defaultViewContainerLocation": true,
    "viewContainer.workbench.view.editSessions.enabled": false,
    "treeView.workbench.views.editSessions.data.enableCollapseAll": true,
    "treeView.workbench.views.editSessions.data.enableRefresh": true,
    "editSessionsCount": 0,
    "workbench.views.editSessions.data.defaultViewLocation": true,
    "workbench.views.editSessions.data.canMove": false,
    "workspacePlatform": "windows",
    "hasGallery": true,
    "hasLocalServer": true,
    "hasRemoteServer": false,
    "hasWebServer": false,
    "workbench.views.extensions.installed.defaultViewLocation": true,
    "workbench.views.extensions.popular.defaultViewLocation": true,
    "extensions.recommendedList.defaultViewLocation": true,
    "workbench.views.extensions.enabled.defaultViewLocation": true,
    "workbench.views.extensions.disabled.defaultViewLocation": true,
    "workbench.views.extensions.marketplace.defaultViewLocation": true,
    "workbench.views.extensions.searchInstalled.defaultViewLocation": true,
    "workbench.views.extensions.searchRecentlyUpdated.defaultViewLocation": true,
    "workbench.views.extensions.searchEnabled.defaultViewLocation": true,
    "workbench.views.extensions.searchDisabled.defaultViewLocation": true,
    "workbench.views.extensions.searchOutdated.defaultViewLocation": true,
    "workbench.views.extensions.searchBuiltin.defaultViewLocation": true,
    "workbench.views.extensions.searchWorkspaceUnsupported.defaultViewLocation": true,
    "workbench.views.extensions.workspaceRecommendations.defaultViewLocation": true,
    "workbench.views.extensions.otherRecommendations.defaultViewLocation": true,
    "workbench.views.extensions.builtinFeatureExtensions.defaultViewLocation": true,
    "workbench.views.extensions.builtinThemeExtensions.defaultViewLocation": true,
    "workbench.views.extensions.builtinProgrammingLanguageExtensions.defaultViewLocation": true,
    "workbench.views.extensions.untrustedUnsupportedExtensions.defaultViewLocation": true,
    "workbench.views.extensions.untrustedPartiallySupportedExtensions.defaultViewLocation": true,
    "workbench.views.extensions.virtualUnsupportedExtensions.defaultViewLocation": true,
    "workbench.views.extensions.virtualPartiallySupportedExtensions.defaultViewLocation": true,
    "workbench.views.extensions.deprecatedExtensions.defaultViewLocation": true,
    "workbench.views.extensions.installed.canMove": false,
    "workbench.views.extensions.popular.canMove": false,
    "extensions.recommendedList.canMove": false,
    "workbench.views.extensions.enabled.canMove": false,
    "workbench.views.extensions.disabled.canMove": false,
    "workbench.views.extensions.marketplace.canMove": false,
    "workbench.views.extensions.searchInstalled.canMove": false,
    "workbench.views.extensions.searchRecentlyUpdated.canMove": false,
    "workbench.views.extensions.searchEnabled.canMove": false,
    "workbench.views.extensions.searchDisabled.canMove": false,
    "workbench.views.extensions.searchOutdated.canMove": false,
    "workbench.views.extensions.searchBuiltin.canMove": false,
    "workbench.views.extensions.searchWorkspaceUnsupported.canMove": false,
    "workbench.views.extensions.workspaceRecommendations.canMove": false,
    "workbench.views.extensions.otherRecommendations.canMove": false,
    "workbench.views.extensions.builtinFeatureExtensions.canMove": false,
    "workbench.views.extensions.builtinThemeExtensions.canMove": false,
    "workbench.views.extensions.builtinProgrammingLanguageExtensions.canMove": false,
    "workbench.views.extensions.untrustedUnsupportedExtensions.canMove": false,
    "workbench.views.extensions.untrustedPartiallySupportedExtensions.canMove": false,
    "workbench.views.extensions.virtualUnsupportedExtensions.canMove": false,
    "workbench.views.extensions.virtualPartiallySupportedExtensions.canMove": false,
    "workbench.views.extensions.deprecatedExtensions.canMove": false,
    "isWorkspaceTrustEnabled": true,
    "isWorkspaceTrusted": true,
    "inCommandsPicker": false,
    "workbench.accounts.experimental.showEntitlements": true,
    "languageSupportsDisassembleRequest": false,
    "mergeConflictsCount": 0,
    "resource": "vscode-settings:settingseditor",
    "resourceScheme": "vscode-settings",
    "resourceFilename": "settingseditor",
    "resourceDirname": ".",
    "resourcePath": "settingseditor",
    "resourceLangId": "unknown",
    "resourceExtname": "",
    "resourceSet": true,
    "isFileSystemResource": false,
    "isAuxiliaryEditorPart": false,
    "editorPartMultipleEditorGroups": false,
    "editorPartMaximizedEditorGroup": false,
    "activeEditorIsDirty": false,
    "activeEditorIsNotPreview": false,
    "activeEditorIsFirstInGroup": true,
    "activeEditorIsLastInGroup": false,
    "activeEditorIsPinned": false,
    "groupEditorsCount": 2,
    "editorPinnedAndUnpinnedTabsVisible": false,
    "breadcrumbsPossible": false,
    "breadcrumbsVisible": false,
    "breadcrumbsActive": false,
    "inSettingsEditor": true,
    "inSettingsSearch": false,
    "settingsTocRowFocus": false,
    "settingRowFocus": true,
    "listFocus": true,
    "listScrollAtBoundary": "both",
    "listSupportsMultiselect": false,
    "listSelectionNavigation": false,
    "listSupportsFind": false,
    "listHasSelectionOrFocus": false,
    "listDoubleSelection": false,
    "listMultiSelection": false,
    "treeElementCanCollapse": false,
    "treeElementHasParent": false,
    "treeElementCanExpand": false,
    "treeElementHasChild": false,
    "treeFindOpen": false
}
```

**Comment by nhedger** [comment]
@Malix-off do you still intend to make this work, or should we close and revisit at a later time?

**Comment by Malix-off** [comment]
> @Malix-off do you still intend to make this work, or should we close and revisit at a later time?

I originally intended to deep dive into VSCode docs, but found some even more problems in their docs and API, and now have other more important problems on the line.

I've converted this PR to a draft, until I find another solution.

If a commit fixes the same concern, I will make sure to close it.

Edits are allowed and welcomed from maintainers and contributors if a better solution has been found.

**Comment by Malix-off** [comment]
> The following wouldn't work anymore, for example:
>
> - `./biome`
> - `/some/path/with/biome`

Have you tested them?
They seem to be valid URI in theory

File: package.json [hunk 1]
```diff
@@ -83,6 +83,7 @@
 				},
 				"biome.lspBin": {
 					"type": "string",
```
Comment (Malix-off) [inline package.json:85]:
```suggestion
					"type": ["string", "null"],
```

File: package.json [hunk 2]
```diff
@@ -83,6 +83,7 @@
...
+					"format": "uri",
 					"default": "",
```
Comment (Malix-off) [inline package.json:87]:
```suggestion
					"default": null,
```

**Comment by Malix-off** [comment]
Yep, relative paths doesn't work

**Comment by Malix-off** [comment]
I'm going to try pattern

**Comment by Malix-off** [comment]
made it kinda work with pattern, but could cause more problem that it would solve, closing

**Pull request closed without merging**